multicluster/controllers/multicluster/label_identity_controller.go

```diff
@@ -0,0 +1,340 @@
+				Namespace: o.GetNamespace(),
+			},
+		},
+	}
+}
+
+// onPodDelete removes the Pod and label identity relation from the cache, and updates the cluster's
+// label identity ResourceExport if necessary (the Pod deletion event causes some label identities
+// to no longer being present in the cluster).
+func (r *LabelIdentityReconciler) onPodDelete(ctx context.Context, podKey, normalizedLabel string) error {
+	if resExportNeedUpdate := r.deleteLabelMapping(podKey, normalizedLabel); resExportNeedUpdate {
+		labelsInCluster := r.getLabelsInCluster()
+		if err := r.refreshLabelIdentityResourceExport(ctx, labelsInCluster); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *LabelIdentityReconciler) deleteLabelMapping(podKey, normalizedLabel string) bool {
+	r.labelMutex.Lock()
+	defer r.labelMutex.Unlock()
+
+	originalLabel, ok := r.podLabelCache[podKey]
+	if !ok {
+		// Might be a host-network Pod deletion
+		return false
+	} else if normalizedLabel == "" {
+		// Use the cached label identity for the Pod to clean up labelToPodsCache
+		normalizedLabel = originalLabel
+	}
+	delete(r.podLabelCache, podKey)
+	podNames, ok := r.labelToPodsCache[normalizedLabel]
+	if !ok || !podNames.Has(podKey) {
+		klog.V(2).InfoS("Label is already removed from cache", "Pod", podKey, "label", normalizedLabel)
+		return false
+	}
+	podNames.Delete(podKey)
+	if len(podNames) == 0 {
+		klog.V(2).InfoS("Label no longer exist in the cluster", "label", normalizedLabel)
+		delete(r.labelToPodsCache, normalizedLabel)
+		return true
+	}
+	return false
+}
+
+func (r *LabelIdentityReconciler) checkPodLabelCache(pod, normalizedLabel string) (string, bool) {
+	r.labelMutex.RLock()
+	defer r.labelMutex.RUnlock()
+
+	cachedPodLabel, labelChanged := "", false
+	cachedPodLabel, ok := r.podLabelCache[pod]
+	if !ok {
+		labelChanged = true
+	} else if cachedPodLabel != normalizedLabel {
+		labelChanged = true
+	}
+	return cachedPodLabel, labelChanged
+}
+
+// onPodUpdate updates the Pod and label identity relation in the cache, and updates the cluster's
+// label identity ResourceExport if necessary (the Pod creation event causes some new label identities
+// to appear in the cluster).
+func (r *LabelIdentityReconciler) onPodUpdate(ctx context.Context, podKey, normalizedLabel string) error {
+	if resExportNeedUpdate := r.updateLabelMapping(podKey, normalizedLabel); resExportNeedUpdate {
+		labelsInCluster := r.getLabelsInCluster()
+		if err := r.refreshLabelIdentityResourceExport(ctx, labelsInCluster); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// onPodUpdates updates Pod and label identity relations in batch, and updates the cluster's label identity
+// ResourceExport if ncessary.
+func (r *LabelIdentityReconciler) onPodUpdates(ctx context.Context, labelUpdates map[string]string) error {
+	resExportNeedUpdate := false
+	for podKey, normalizedLabel := range labelUpdates {
+		if updated := r.updateLabelMapping(podKey, normalizedLabel); updated {
```

Comment:
Looks like this can be called only when Namespace is added or label is changed,  I think you don't need `resExportNeedUpdate`, it should always update the ResourceExport.

Reply:
This is not the case. Consider Pods A, B, C with labels L1, L1 and L2 initially. If Pod B changes its label to L2. the ResourceExport does not need to be updated since the labels in the cluster remain to be [L1, L2]